### PR TITLE
286 grid rewrite

### PR DESCRIPTION
**Pull request by CamBuscher**
This is the current project state. 

The deployed version has both the existing app on the top half of the page, and the second half of the page is the new version. 

This was done in order to provide access to some functionality from the existing app that hasn't been creating for the new app, namely the time slot & generate grid modals. It does create some CSS nastiness with the fixed sticky headers overlapping on scroll.

### CODE REVIEW

File: app/javascript/apiCalls.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+      method: "PATCH",
+      headers: {
+        "Content-Type": "application/json",
+        "X-CSRF-Token": csrfToken
+      },
+      credentials: "same-origin",
+      body: data
+    })
+  }
+}
```
Comment (TimothyClayton) [inline app/javascript/apiCalls.js]: That red 'Ghostbuster' and return symbol there is no new line at end of file. It's good style/practice to have a new line at the end of files, and fortunately the text editor can handle that for you. I think you both use VS Code? First answer below should work, though you may have to open the file and save it. You can scroll through the files in this PR see which ones have no new line. 

https://stackoverflow.com/questions/44704968/visual-studio-code-insert-new-line-at-the-end-of-files

File: app/assets/stylesheets/modules/_schedule.scss [hunk 2]
```diff
@@ -383,3 +383,192 @@ input[type=number]::-webkit-outer-spin-button {
...
+  padding-left: 0px;
+  display: inline-block;
+}
+
+.ruler_tick {
+  height: $pixel-step;
+  text-align: right;
+  position: relative;
+  width: 48px;
+}
```
Comment (TimothyClayton) [inline app/assets/stylesheets/modules/_schedule.scss:472]: New line here for consistency

File: app/assets/stylesheets/modules/_schedule.scss [hunk 3]
```diff
@@ -383,3 +383,192 @@ input[type=number]::-webkit-outer-spin-button {
   background-color: white;
   z-index: 1000;
 }
+
```
Comment (TimothyClayton) [inline app/assets/stylesheets/modules/_schedule.scss:386]: I like the nesting here. It doesn't look like it goes more than 4 deep, and only in a couple places. It's good to keep the nesting minimized (I usually shoot for max 3), as the autogenerated selectors can get unwieldy and complicate using dev tools.

File: app/assets/javascripts/staff/program/grid.js [hunk 4]
```diff
@@ -5,7 +5,7 @@
   if (typeof(window.Schedule.Grid) !== 'undefined') {
     return window.Schedule.Grid;
   }
-
+
```
Comment (TimothyClayton) [inline app/assets/javascripts/staff/program/grid.js:8]: Looks like just a straggling space here to remove...I think. Not sure if that's only on the github side PR view.

File: app/assets/javascripts/staff/program/grid.js [hunk 5]
```diff
@@ -190,7 +190,7 @@
     if (url == null || url.length==0) {
       return;
     }
-
+    console.log('url:', url)
```
Comment (TimothyClayton) [inline app/assets/javascripts/staff/program/grid.js]: Generally, remove `console.log`'s unless it's a case where something needs debugging on prod.

File: app/javascript/apiCalls.js (hunk 1 above)
Comment (TimothyClayton) [inline app/javascript/apiCalls.js]: That red 'Ghostbuster' and return symbol are 'No new line at end of file' indicators. It's good style/practice to have a new line at the end of files, and fortunately your text editor can automate that so you don't have to think about it. I think you both use VS Code? If so, the first answer in the link below should suffice. You may have to open and save each file that has no new line, which you can see by scrolling through the PR.

https://stackoverflow.com/questions/44704968/visual-studio-code-insert-new-line-at-the-end-of-files

File: app/javascript/components/Schedule.js [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+            endTime={endTime}
+            ripTime={this.ripTime}
+            changeDragged={this.changeDragged}
+            draggedSession={draggedSession}
+            csrf={csrf}
+            sessions={sessions}
+            scheduleSession={this.scheduleSession}
+            tracks={tracks}
+          />
+          <UnschedledArea
```
Comment (TimothyClayton) [inline app/javascript/components/Schedule.js]: `UnschedledArea` ->`UnscheduledArea` (missing 'u')

**Review by TimothyClayton — CHANGES_REQUESTED** [review]
Comments should be inline; a few changes requested. Plus what Marty indicated about removing the old grid. I also noticed a minor UI thing but wasn't sure where in the code to comment. When I drag over a grid item, the background becomes transparent, which I thought seemed odd but not a show stopper.

File: app/assets/stylesheets/modules/_schedule.scss (hunk 3 above)
Comment (CamBuscher) [reply app/assets/stylesheets/modules/_schedule.scss:386]: Thank you for this, I will definitely limit nesting in the future!

File: app/javascript/components/Schedule.js (hunk 6 above)
Comment (CamBuscher) [reply app/javascript/components/Schedule.js]: I'm impressed this didn't cause me issues. Good eye.

File: app/javascript/apiCalls.js (hunk 1 above)
Comment (CamBuscher) [reply app/javascript/apiCalls.js]: This solution was awesome @gavin-love

**Review by TimothyClayton — COMMENTED** [review]
Looks good to me. I'll defer to Marty's feedback and any needed updates before approving.

**Review by TimothyClayton — COMMENTED** [review]
Removing 'Request Changes' state of PR, and Github is forcing me to leave a comment 😬. Sheesh that didn't work. See comment above.

**Comment by CamBuscher** [comment]
Added following CSS changes:

- Horizontal scroll from any vertical position when needed.
- Unscheduled area always shows on right side of grid, overlapping grid as necessary.
- Headers are sticky as required, also accommodate for collapsing nav bars when window width shrinks.
- "Ruler" with times on left side is also always visible. Let me know what you think of the opacity on it. 

Also added functionality to the searching of unscheduled area sessions, allowing for search by name & track. 
Added clear button for unscheduled search.

**Review by TimothyClayton — APPROVED** [review]

**Pull request merged**